core/retries/src/main/java/software/amazon/awssdk/retries/internal/ratelimiter/RateLimiterTokenBucket.java

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.retries.internal.ratelimiter;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * The {@link RateLimiterTokenBucket} keeps track of past throttling responses and adapts to slow down the send rate to adapt to
+ * the service. It does this by suggesting a delay amount as result of a {@link #tryAcquire()} call. Callers must update its
+ * internal state by calling {@link #updateRateAfterThrottling()} when getting a throttling response or
+ * {@link #updateRateAfterSuccess()} when getting successful response.
+ *
+ * <p>The algorithm used is adapted from the network congestion avoidance algorithm
+ * <a href="https://en.wikipedia.org/wiki/CUBIC_TCP">CUBIC</a>.
+ */
+@SdkInternalApi
+public class RateLimiterTokenBucket {
+    private final AtomicReference<PersistentState> stateReference;
+    private final RateLimiterClock clock;
+
+    RateLimiterTokenBucket(RateLimiterClock clock) {
+        this.clock = clock;
+        this.stateReference = new AtomicReference<>(new PersistentState());
+    }
+
+    /**
+     * Acquires one token from the rate limiter and returns a response with a delay value (possibly zero). Callers are expected to
+     * wait for at least this amount of time before retrying the request. See {@link RateLimiterAcquireResponse#delay()}.
+     */
+    public RateLimiterAcquireResponse tryAcquire() {
+        StateUpdate<Duration> update = updateState(ts -> ts.tokenBucketAcquire(clock, 1.0));
+        return RateLimiterAcquireResponse.create(update.result);
+    }
+
+    /**
+     * Updates the estimated send rate after a throttling response.
+     */
+    public RateLimiterUpdateResponse updateRateAfterThrottling() {
+        StateUpdate<Void> update = consumeState(ts -> ts.updateClientSendingRate(clock, true));
+        return RateLimiterUpdateResponse.builder()
+            .measuredTxRate(update.newState.measuredTxRate())
+            .fillRate(update.newState.fillRate())
+            .build();
+    }
+
+    /**
+     * Updates the estimated send rate after a successful response.
+     */
+    public RateLimiterUpdateResponse updateRateAfterSuccess() {
```

Comment:
I see that the implementation at https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/retry/RateLimitingTokenBucket.java uses commonly accepted API naming conventions, such as refill() as direct API of the bucket

Reply:
Yeah, I have seen that as well. This class encapsulates that logic and exposes a better naming which makes the reading of the code using it easier to understand.

Reply:
I have added some more javadocs where appropriate, I didn't copy the pattern of adding a javadoc with the pseudo code below the actual java implementation since I don't think that adds value.